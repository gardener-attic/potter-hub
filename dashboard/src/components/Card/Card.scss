.Card,
.Card__inner {
  height: 100%;
  width: 100%;
}

.Card {
  margin: 0 1% 1.5em;
}

.Card a,
.Card a:hover {
  // TODO: See why we need important here - https://github.com/bitnami/hex/issues/217
  // sass-lint:disable no-important
  text-decoration: none !important;
}

.Card__Icon img {
  max-height: 64px;
}

/* max 4 items per row (default) */
@media screen and (min-width: 480px) {
  .Card.Card-responsive {
    width: 48%;
  }
}

@media screen and (min-width: 768px) {
  .Card.Card-responsive {
    width: 31%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive {
    width: 23%;
  }
}

/* max 3 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-3 {
    width: 48%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive-3 {
    width: 31%;
  }
}

/* max 2 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-2 {
    width: 48%;
  }
}
