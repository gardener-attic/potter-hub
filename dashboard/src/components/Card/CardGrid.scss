.CardGrid {
  display: flex;
  margin: 0 -1%;
  justify-content: space-between;
  flex-wrap: wrap;
}

.CardGrid:after {
  /* Display the last row properly: https://stackoverflow.com/a/34816625 */
  flex: auto;
  content: "";
}
