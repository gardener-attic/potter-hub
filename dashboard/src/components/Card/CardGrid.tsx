import * as React from "react";

import "./CardGrid.css";

export interface ICardGridProps {
  className?: string;
  children?: React.ReactChildren | React.ReactNode | string;
}

interface ICardGridState {
  scrollEndIndex: number
}

const SCROLL_THRESHOLD_TOP_PIXEL = 120;
const SCROLL_THRESHOLD_BOTTOM_TO_LOAD = 0.8;
const ITEMS_PER_SCROLL = 32

class CardGrid extends React.Component<ICardGridProps, ICardGridState> {
  constructor(props: any) {
    super(props);
    this.state = {
      scrollEndIndex: ITEMS_PER_SCROLL,
    };
  }

  public componentDidMount() {
    window.addEventListener("scroll", this.onScroll);
  }

  public componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll)
  }

  public render() {
    return (
      <div className={`CardGrid padding-v-big ${this.props.className || ""}`}>
        {React.Children.toArray(this.props.children).slice(0, this.state.scrollEndIndex)}
      </div>
    );
  }

  private onScroll = (): void => {
    // calculates, if the current scroll position further than the SCROLL_THRESHOLD_BOTTOM_TO_LOAD threshold
    const isScrollAtBottom = window.innerHeight + document.documentElement.scrollTop >
      document.documentElement.offsetHeight * SCROLL_THRESHOLD_BOTTOM_TO_LOAD;

    // calculates, if the current scroll position is above the SCROLL_THRESHOLD_TOP_PIXEL
    const isScrollAtTop = document.documentElement.scrollTop < SCROLL_THRESHOLD_TOP_PIXEL;

    if (isScrollAtBottom && this.state.scrollEndIndex < React.Children.count(this.props.children)) {
      this.setState((state) => ({
        scrollEndIndex: state.scrollEndIndex + ITEMS_PER_SCROLL
      }))
    }

    if (isScrollAtTop && this.state.scrollEndIndex > ITEMS_PER_SCROLL) {
      this.setState((state) => ({
        scrollEndIndex: ITEMS_PER_SCROLL
      }))
    }
  }
}

export default CardGrid;
