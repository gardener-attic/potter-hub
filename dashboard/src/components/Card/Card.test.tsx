import { shallow } from "enzyme";
import * as React from "react";

import Card from "./Card";

describe("cssClass", () => {
  it("should render with the default class", () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.find(".Card").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("should use the className property", () => {
    const wrapper = shallow(<Card className="foo" />);
    expect(wrapper.find(".foo").exists()).toBe(true);
  });

  it("should add the responsive class", () => {
    const wrapper = shallow(<Card responsive={true} />);
    expect(wrapper.find(".Card-responsive").exists()).toBe(true);
  });

  it("should add the responsive class with colums", () => {
    const wrapper = shallow(<Card responsive={true} responsiveColumns={1} />);
    expect(wrapper.find(".Card-responsive-1").exists()).toBe(true);
  });
});

it("should render the children elements", () => {
  const wrapper = shallow(
    <Card>
      <div>foo</div>
    </Card>,
  );
  expect(wrapper.text()).toContain("foo");
});
