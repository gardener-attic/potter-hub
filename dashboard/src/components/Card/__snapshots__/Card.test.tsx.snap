// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssClass should render with the default class 1`] = `
<article
  className="Card "
>
  <div
    className="Card__inner bg-white elevation-1"
  />
</article>
`;
